hack/update-workspace-mirror.sh

```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+# Copyright 2018 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+
+cd "${KUBE_ROOT}"
+
+BUCKET=gs://k8s-bazel-cache
+
+gsutil acl get "${BUCKET}" > /dev/null
+
+ETCD_VERSION=$(grep -m1 ETCD_VERSION WORKSPACE | grep -o '[0-9.]*')
+
+for url in $(grep 'mirror(' WORKSPACE | grep -o 'https://[^"]*'); do
+  url=$(echo "${url}" | sed -e "s/%s/${ETCD_VERSION}/g")
```

Comment:
this one is a little unfortunate. :\

Reply:
the alternative hacks are probably worse :/

Reply:
Let bazel process the file for you:
`WORKSPACE`
```
load("//somewhere:something.bzl", "foo")
foo()
```
`bash$ bazel run //:update_http_archive_cache`
`somewhere/something.bzl`
```
def foo():
  urls = []
  for k, v in native.existing_rules().items():
    if v["kind"] in ["http_archive", "new_http_archive"]:
      print("MATCH", k, "url", v.get("url"), "urls", v.get("urls"))
      u = v.get("url")
      us = v.get("urls")
      if u:
        urls.append(u)
      if us:
        urls.extend(us)
    else:
      # print("no", k)
      pass
  if urls:
    print("lets update", urls)
  fail("demo over!")
```

https://docs.bazel.build/versions/master/skylark/macros.html
https://docs.bazel.build/versions/master/skylark/lib/native.html#existing_rules

Reply:
Move `foo()` into a `bar/BUILD` and change `foo()` to generate a binary rule:

`bar/BUILD.bazel`
```
load("somewhere/something.bzl", "foo")
foo("update-cache")
```

`something.bzl`
```
def foo(name):
  ...
  if urls:
    print("lets update", urls)
    native.sh_binary(
      name = name,
      srcs = ["cache.sh"],  # the curl | gsutil line in this file
      args = urls,
    )
```

Then `bazel run //bar:update-cache`

https://docs.bazel.build/versions/master/skylark/cookbook.html#macro_native
https://docs.bazel.build/versions/master/be/shell.html#sh_binary

Reply:
Create a [genrule](https://docs.bazel.build/versions/master/be/general.html#genrule) that takes as input the list of URLs we use and outputs the list of URLs that don't have a cached version. 

 
`foo.bzl`
```
def cached_url(u):
  return PREFIX + u

def foofunc():
  urls = ... # see above
  native.genrule(
    name="blah",
    outs=["missing-cache.urls"],
    cmd="""
urls=( {urls} )
touch $@
for u in urls; do
  curl --fail --head $u || echo $u >> $@
done
  """.format(urls=" ".join([cached_url(u) for u in urls]))
```


`bar/BUILD`
```
load(":foo.bzl", "foofunc")
foofunc() # generates the blah rule that generates the file
```

`bazel build //bar:blah && cat bazel-genfiles/bar/missing-cache.urls`

Reply:
Make an `sh_test()` rule which ensure that either a) there are no uncached urls or b) it can successfully call the update_cache rule to update them

```
load("//somewhere:something.bzl", "foo")
foofunc()
foo("update-cache")

sh_test(
  name = "verify_cache_test",
  srcs = ["verify_cache_test.sh"], # [[ -z "$(cat $1)" ]] && exit 0 || $2 <<<$1 
  args = ["$(location :missing-cache.urls)", "$(location :update-cache)"],
  local = True, # use user gsutil creds
)
```

https://docs.bazel.build/versions/master/be/shell.html#sh_test

Reply:
If we do all of this then whenever someone changes the WORKSPACE file presubmits tests will fail unless the author (or someone else with write access to the cache, hopefully not the presubmit service account) has run the update_cache binary with the new url list.

All of this (or something like it) would be most useful across time, but all is not necessary for this PR.

Can we at least do something so bazel is the one processing the urls rather than `grep` and `sed`?